dataRetriever/resolvers/epochproviders/arithmeticEpochProvider.go

```diff
@@ -0,0 +1,121 @@
+	RoundTimeInMilliseconds uint64
+	StartTime               int64
+}
+
+type arithmeticEpochProvider struct {
+	sync.RWMutex
+	currentComputedEpoch       uint32
+	headerEpoch                uint32
+	headerTimestampForNewEpoch uint64
+	roundsPerEpoch             uint32
+	roundTimeInMilliseconds    uint64
+	startTime                  int64
+	getUnixHandler             func() int64
+}
+
+// NewArithmeticEpochProvider returns a new arithmetic epoch provider able to mathematically compute the current network epoch
+// based on the last block saved, considering the block's timestamp and epoch in respect with the current time
+func NewArithmeticEpochProvider(arg ArgArithmeticEpochProvider) (*arithmeticEpochProvider, error) {
+	if arg.RoundsPerEpoch == 0 {
+		return nil, ErrInvalidRoundsPerEpoch
+	}
+	if arg.RoundTimeInMilliseconds == 0 {
+		return nil, ErrInvalidRoundTimeInMilliseconds
+	}
+	if arg.StartTime < 0 {
+		return nil, ErrInvalidStartTime
+	}
+	aep := &arithmeticEpochProvider{
+		headerEpoch:                0,
+		headerTimestampForNewEpoch: uint64(arg.StartTime),
+		roundsPerEpoch:             arg.RoundsPerEpoch,
+		roundTimeInMilliseconds:    arg.RoundTimeInMilliseconds,
+		startTime:                  arg.StartTime,
+	}
+	aep.getUnixHandler = func() int64 {
+		return time.Now().Unix()
+	}
+	aep.computeCurrentEpoch() //based on the genesis provided data
+
+	return aep, nil
+}
+
+// EpochIsActiveInNetwork returns true if the persister for the given epoch is active in the network
+func (aep *arithmeticEpochProvider) EpochIsActiveInNetwork(epoch uint32) bool {
+	aep.RLock()
+	defer aep.RUnlock()
+
+	if aep.currentComputedEpoch < epoch {
+		return true
+	}
+
+	return aep.currentComputedEpoch-epoch <= deltaEpochActive
+}
+
+// EpochStartAction is called whenever the epoch was changed and will cause the re-computation of the current epoch
+func (aep *arithmeticEpochProvider) EpochStartAction(header data.HeaderHandler) {
+	if check.IfNil(header) {
+		return
+	}
+
+	aep.Lock()
+	aep.headerEpoch = header.GetEpoch()
+	aep.headerTimestampForNewEpoch = header.GetTimeStamp()
+
+	aep.computeCurrentEpoch()
+	aep.Unlock()
+}
+
+// EpochStartPrepare does nothing
+func (aep *arithmeticEpochProvider) EpochStartPrepare(_ data.HeaderHandler, _ data.BodyHandler) {
+}
+
+// NotifyOrder will return the core.CurrentNetworkEpochProvider value
+func (aep *arithmeticEpochProvider) NotifyOrder() uint32 {
+	return core.CurrentNetworkEpochProvider
+}
+
+func (aep *arithmeticEpochProvider) computeCurrentEpoch() {
+	currentTimeStamp := uint64(aep.getUnixHandler())
+
+	if currentTimeStamp < aep.headerTimestampForNewEpoch {
+		aep.currentComputedEpoch = aep.headerEpoch
+		return
+	}
+
+	diffTimeStampInSeconds := currentTimeStamp - aep.headerTimestampForNewEpoch
+	diffTimeStampInMilliseconds := diffTimeStampInSeconds * millisecondsInOneSecond
+	diffRounds := diffTimeStampInMilliseconds / aep.roundTimeInMilliseconds
+	diffEpochs := diffRounds / uint64(aep.roundsPerEpoch+1)
```

Comment:
why +1?
I have written the next unit test:
``` golang
func TestEpochsDiff(t *testing.T) {
	roundTimeInMilliseconds := uint64(5000)
	roundsPerEpoch := 500
	millisInASec := 1000

	currentTimeStamp := uint64(1600000000)

	epochDuration := uint64(roundsPerEpoch) * (roundTimeInMilliseconds/uint64(millisInASec))

	headerTimestampForNewEpoch := currentTimeStamp - (epochDuration + 5)

	diffTimeStampInSeconds := currentTimeStamp - headerTimestampForNewEpoch
	diffTimeStampInMilliseconds := diffTimeStampInSeconds * 1000
	diffRounds := diffTimeStampInMilliseconds / roundTimeInMilliseconds
	diffEpochs := diffRounds / uint64(roundsPerEpoch+1)

	fmt.Println(diffEpochs)
}
```

on this line:
`headerTimestampForNewEpoch := currentTimeStamp - (epochDuration + 5)`
if you add 5, the epochs diff is `1` but for 4 the epochs diff will be `0`. so for one second difference, the result is quite different.

Reply:
I remember this: we have an extra round per epoch in each epoch. 
So this line from your test:
```GO
epochDuration := uint64(roundsPerEpoch) * (roundTimeInMilliseconds/uint64(millisInASec))
```
should be changed to 
```GO
epochDuration := uint64(roundsPerEpoch + 1) * (roundTimeInMilliseconds/uint64(millisInASec))
```